File: src/foundry/agents/cursor.py
"""Cursor coding-agent providers.

Two ways to hand approved work to Cursor, both implementing
:class:`CodingAgentProvider`:

- :class:`CursorViaLinearProvider` - the *preferred* path. Foundry has already
  gathered context, classified risk and obtained approval; it then delegates to
  Cursor through Linear by posting an ``@Cursor`` comment with the governed
  instructions. Cursor's Linear integration runs the cloud agent, reports status
  back in Linear and auto-opens the PR. Foundry stays the control plane above.
  (Requires the Cursor⨉Linear integration: a Cursor admin connection + GitHub.)

- :class:`CursorCloudAgentProvider` - the direct Cloud Agents API
  (``POST https://api.cursor.com/v0/agents``). Useful when there is no Linear
  hand-off (e.g. a non-Linear trigger). The HTTP calls are injected so this is
  testable without network and without a real Cursor key.

Both go through ``create_job`` on the base class, so the secret-leak guard runs
before anything is dispatched.
"""

from __future__ import annotations

from typing import Any, Callable

from foundry.connectors.base import IssueTracker
from foundry.connectors.comments import format_cursor_delegation
from foundry.schemas.agent import (
    CodingAgentJob,
    CodingAgentJobInput,
    CodingAgentJobStatus,
)
from foundry.schemas.common import AgentJobStatus

from .provider import CodingAgentProvider

# Cursor cloud-agent status -> Foundry job status.
_CURSOR_STATUS = {
    "CREATING": AgentJobStatus.CREATED,
    "PENDING": AgentJobStatus.CREATED,
    "RUNNING": AgentJobStatus.RUNNING,
    "FINISHED": AgentJobStatus.SUCCEEDED,
    "COMPLETED": AgentJobStatus.SUCCEEDED,
    "ERROR": AgentJobStatus.FAILED,
    "FAILED": AgentJobStatus.FAILED,
    "CANCELLED": AgentJobStatus.CANCELLED,
    "EXPIRED": AgentJobStatus.CANCELLED,
}


class CursorViaLinearProvider(CodingAgentProvider):
    """Delegate to Cursor by commenting ``@Cursor`` on the Linear issue."""

    name = "cursor_via_linear"

    def __init__(self, tracker: IssueTracker) -> None:
        self._tracker = tracker

    def _dispatch(self, job_input: CodingAgentJobInput) -> CodingAgentJob:
        if not job_input.tracker_issue_id:
            raise ValueError(
                "CursorViaLinearProvider requires job_input.tracker_issue_id "
                "(the Linear issue to delegate on)"
            )
        body = format_cursor_delegation(job_input.agent_instructions)
        self._tracker.post_comment(job_input.tracker_issue_id, body)
        # The agent now runs inside Cursor's Linear integration; progress and the
        # PR arrive via Linear/GitHub webhooks, which drive orchestrator.record_pr.
        return CodingAgentJob(
            job_id=f"cursor-linear:{job_input.tracker_issue_id}",
            provider=self.name,
            status=AgentJobStatus.RUNNING,
        )

    def get_job_status(self, job_id: str) -> CodingAgentJobStatus:
        # Status is observed out-of-band (Linear/GitHub), not polled here.
        return CodingAgentJobStatus(
            job_id=job_id, provider=self.name, status=AgentJobStatus.RUNNING
        )

    def cancel_job(self, job_id: str) -> None:  # pragma: no cover - no-op
        # Cancellation is a human action in Linear/Cursor for this path.
        return None


class CursorCloudAgentProvider(CodingAgentProvider):
    """Launch a Cursor cloud agent directly via the Cloud Agents API."""

    name = "cursor_cloud"
    _AGENTS_URL = "https://api.cursor.com/v0/agents"

    def __init__(
        self,
        *,
        http_post: Callable[[str, dict, dict], dict],
        http_get: Callable[[str, dict], dict] | None = None,
        auto_create_pr: bool = True,
    ) -> None:
        # http_post(url, json_body, headers) -> response json
        # http_get(url, headers) -> response json
        self._http_post = http_post
        self._http_get = http_get
        self._auto_create_pr = auto_create_pr

    def _dispatch(self, job_input: CodingAgentJobInput) -> CodingAgentJob:
        body = {
            "prompt": {"text": job_input.agent_instructions},
            "source": {
                "repository": self._repo_url(job_input.repo),
                "ref": job_input.base_branch,
            },
            "target": {
                "autoCreatePr": self._auto_create_pr,
                "branchName": job_input.branch_name,
            },
        }
        # The API key is supplied by the injected transport's headers, never in
        # the job input - so the secret guard never sees it.
        response = self._http_post(self._AGENTS_URL, body, {})
        return CodingAgentJob(
            job_id=str(response["id"]),
            provider=self.name,
            status=_CURSOR_STATUS.get(response.get("status", ""), AgentJobStatus.CREATED),
        )

    def get_job_status(self, job_id: str) -> CodingAgentJobStatus:
        if self._http_get is None:  # pragma: no cover - requires injected client
            raise RuntimeError("CursorCloudAgentProvider needs http_get to poll status")
        data = self._http_get(f"{self._AGENTS_URL}/{job_id}", {})
        target = data.get("target", {}) or {}
        return CodingAgentJobStatus(
            job_id=job_id,
            provider=self.name,
            status=_CURSOR_STATUS.get(data.get("status", ""), AgentJobStatus.RUNNING),
            branch=target.get("branchName"),
            pr_url=target.get("prUrl") or target.get("url"),
        )

    def cancel_job(self, job_id: str) -> None:
        self._http_post(f"{self._AGENTS_URL}/{job_id}/cancel", {}, {})

    @staticmethod
    def _repo_url(repo: str) -> str:
        if repo.startswith("http://") or repo.startswith("https://"):
            return repo
        return f"https://github.com/{repo}"

File: dotnet/src/Foundry/Agents/Cursor.cs
// Cursor coding-agent providers.
//
// Two ways to hand approved work to Cursor, both extending CodingAgentProvider:
//
// - CursorViaLinearProvider - the *preferred* path. Foundry has already
//   gathered context, classified risk and obtained approval; it then delegates
//   to Cursor through Linear by posting an @Cursor comment with the governed
//   instructions. Cursor's Linear integration runs the cloud agent, reports
//   status back in Linear and auto-opens the PR. Foundry stays the control
//   plane above.
//
// - CursorCloudAgentProvider - the direct Cloud Agents API
//   (POST https://api.cursor.com/v0/agents). Useful when there is no Linear
//   hand-off. The HTTP calls are injected so this is testable without network
//   and without a real Cursor key.
//
// Both go through CreateJob on the base class, so the secret-leak guard runs
// before anything is dispatched.

using System.Text.Json;
using Foundry.Connectors;
using Foundry.Schemas;

namespace Foundry.Agents;

/// <summary>httpGet(url, headers) -> response json.</summary>
public delegate JsonElement JsonHttpGet(string url, IReadOnlyDictionary<string, string> headers);

/// <summary>Delegate to Cursor by commenting @Cursor on the Linear issue.</summary>
public sealed class CursorViaLinearProvider : CodingAgentProvider
{
    public const string ProviderName = "cursor_via_linear";

    private readonly IIssueTracker _tracker;

    public CursorViaLinearProvider(IIssueTracker tracker)
    {
        _tracker = tracker;
    }

    public override string Name => ProviderName;

    protected override CodingAgentJob Dispatch(CodingAgentJobInput jobInput)
    {
        if (string.IsNullOrEmpty(jobInput.TrackerIssueId))
        {
            throw new ArgumentException(
                "CursorViaLinearProvider requires jobInput.TrackerIssueId (the Linear issue to delegate on)");
        }
        var body = Comments.FormatCursorDelegation(jobInput.AgentInstructions);
        _tracker.PostComment(jobInput.TrackerIssueId, body);
        // The agent now runs inside Cursor's Linear integration; progress and
        // the PR arrive via Linear/GitHub webhooks, which drive the
        // orchestrator's RecordPr.
        return new CodingAgentJob
        {
            JobId = $"cursor-linear:{jobInput.TrackerIssueId}",
            Provider = Name,
            Status = AgentJobStatus.Running,
        };
    }

    public override CodingAgentJobStatus GetJobStatus(string jobId) =>
        // Status is observed out-of-band (Linear/GitHub), not polled here.
        new() { JobId = jobId, Provider = Name, Status = AgentJobStatus.Running };

    public override void CancelJob(string jobId)
    {
        // Cancellation is a human action in Linear/Cursor for this path.
    }
}

/// <summary>Launch a Cursor cloud agent directly via the Cloud Agents API.</summary>
public sealed class CursorCloudAgentProvider : CodingAgentProvider
{
    public const string ProviderName = "cursor_cloud";
    private const string AgentsUrl = "https://api.cursor.com/v0/agents";

    // Cursor cloud-agent status -> Foundry job status.
    private static readonly IReadOnlyDictionary<string, AgentJobStatus> CursorStatus =
        new Dictionary<string, AgentJobStatus>
        {
            ["CREATING"] = AgentJobStatus.Created,
            ["PENDING"] = AgentJobStatus.Created,
            ["RUNNING"] = AgentJobStatus.Running,
            ["FINISHED"] = AgentJobStatus.Succeeded,
            ["COMPLETED"] = AgentJobStatus.Succeeded,
            ["ERROR"] = AgentJobStatus.Failed,
            ["FAILED"] = AgentJobStatus.Failed,
            ["CANCELLED"] = AgentJobStatus.Cancelled,
            ["EXPIRED"] = AgentJobStatus.Cancelled,
        };

    private readonly JsonHttpPost _httpPost;
    private readonly JsonHttpGet? _httpGet;
    private readonly bool _autoCreatePr;

    public CursorCloudAgentProvider(
        JsonHttpPost httpPost, JsonHttpGet? httpGet = null, bool autoCreatePr = true)
    {
        _httpPost = httpPost;
        _httpGet = httpGet;
        _autoCreatePr = autoCreatePr;
    }

    public override string Name => ProviderName;

    protected override CodingAgentJob Dispatch(CodingAgentJobInput jobInput)
    {
        var body = JsonSerializer.SerializeToElement(new Dictionary<string, object>
        {
            ["prompt"] = new Dictionary<string, string> { ["text"] = jobInput.AgentInstructions },
            ["source"] = new Dictionary<string, string>
            {
                ["repository"] = RepoUrl(jobInput.Repo),
                ["ref"] = jobInput.BaseBranch,
            },
            ["target"] = new Dictionary<string, object>
            {
                ["autoCreatePr"] = _autoCreatePr,
                ["branchName"] = jobInput.BranchName,
            },
        });
        // The API key is supplied by the injected transport's headers, never
        // in the job input - so the secret guard never sees it.
        var response = _httpPost(AgentsUrl, body, new Dictionary<string, string>())
            ?? throw new InvalidOperationException("Cursor API returned no response body");
        var statusText = response.TryGetProperty("status", out var statusProperty)
            ? statusProperty.GetString() ?? ""
            : "";
        return new CodingAgentJob
        {
            JobId = response.GetProperty("id").ToString(),
            Provider = Name,
            Status = CursorStatus.GetValueOrDefault(statusText, AgentJobStatus.Created),
        };
    }

    public override CodingAgentJobStatus GetJobStatus(string jobId)
    {
        if (_httpGet is null)
        {
            throw new InvalidOperationException("CursorCloudAgentProvider needs httpGet to poll status");
        }
        var data = _httpGet($"{AgentsUrl}/{jobId}", new Dictionary<string, string>());
        JsonElement target = default;
        var hasTarget = data.TryGetProperty("target", out target)
            && target.ValueKind == JsonValueKind.Object;
        var statusText = data.TryGetProperty("status", out var statusProperty)
            ? statusProperty.GetString() ?? ""
            : "";
        return new CodingAgentJobStatus
        {
            JobId = jobId,
            Provider = Name,
            Status = CursorStatus.GetValueOrDefault(statusText, AgentJobStatus.Running),
            Branch = hasTarget ? GetStringOrNull(target, "branchName") : null,
            PrUrl = hasTarget
                ? GetStringOrNull(target, "prUrl") ?? GetStringOrNull(target, "url")
                : null,
            CostUsd = ExtractCost(data),
        };
    }

    public override void CancelJob(string jobId) =>
        _httpPost($"{AgentsUrl}/{jobId}/cancel",
            JsonSerializer.SerializeToElement(new Dictionary<string, object>()),
            new Dictionary<string, string>());

    private static string RepoUrl(string repo) =>
        repo.StartsWith("http://") || repo.StartsWith("https://")
            ? repo
            : $"https://github.com/{repo}";

    private static string? GetStringOrNull(JsonElement element, string property) =>
        element.TryGetProperty(property, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    /// <summary>Provider-reported spend, tolerant of the usage shape evolving.</summary>
    internal static double? ExtractCost(JsonElement data)
    {
        var hasUsage = data.TryGetProperty("usage", out var usage)
            && usage.ValueKind == JsonValueKind.Object;
        foreach (var key in new[] { "totalCostUsd", "costUsd", "totalCost", "cost_usd" })
        {
            JsonElement value = default;
            var found = (hasUsage && usage.TryGetProperty(key, out value))
                || data.TryGetProperty(key, out value);
            if (found && value.ValueKind is not (JsonValueKind.Null or JsonValueKind.Undefined))
            {
                if (value.ValueKind == JsonValueKind.Number)
                {
                    return value.GetDouble();
                }
                if (value.ValueKind == JsonValueKind.String
                    && double.TryParse(value.GetString(), out var parsed))
                {
                    return parsed;
                }
                return null;
            }
        }
        return null;
    }
}
